### DISCUSSION TITLE
How to Integrate ESLint Rules with @monaco-editor/react?

### DISCUSSION THREAD

**Discussion by restarajat** [suren-atoyan/monaco-react#653] [Q&A]
Hi there,

I am currently using the @monaco-editor/react component in my React project to provide a code editor, and I want to add ESLint rules to lint the code as the user types. However, I am unsure how to configure ESLint to work with Monaco Editor in my React app.

What I have tried:
I installed eslint and configured a basic ESLint config (.eslintrc.js).
I tried integrating the eslint-plugin-monaco package, but I couldn't get it to show linting errors in the editor.
I've also tried manually setting up ESLint within Monaco’s API, but I'm unsure about how to make Monaco show the linting results.

My questions:
How can I configure @monaco-editor/react to use my custom ESLint rules?
Are there any packages or plugins that integrate ESLint with Monaco Editor for real-time linting support?
How can I display ESLint errors in the editor with Monaco?

Here’s an example of my current code for setting up Monaco:

`import React, { useRef } from 'react';
import Editor, { loader } from '@monaco-editor/react';
import * as monaco from 'monaco-editor';
import eslint from 'eslint';

loader.config({ monaco });

function CodeEditor() {
    const editorRef = useRef(null);

    // Initialize ESLint instance
    const eslintInstance = new eslint.ESLint({
        baseConfig: {
            parser: '@typescript-eslint/parser', // Ensure using the TypeScript parser
            parserOptions: {
                ecmaVersion: 2020,
                sourceType: 'module',
                ecmaFeatures: {
                    jsx: true,
                },
            },
            rules: {
                'no-console': 'warn', // Add your ESLint rules here
                // other rules...
            },
        },
        useEslintrc: false, // Disable external config loading for simplicity, or set to true to use .eslintrc
    });

    const onEditorMount = (editor, monaco) => {
        editorRef.current = editor;

        // Set up TypeScript support
        monaco.languages.typescript.typescriptDefaults.setDiagnosticsOptions({
            noSemanticValidation: false,
            noSyntaxValidation: false,
        });

        // Custom typings and suggestions for React Native
        monaco.languages.typescript.typescriptDefaults.addExtraLib(
            `declare module 'react-native' { export const View: any; /* add more as needed */ }`,
            'file:///node_modules/@types/react-native/index.d.ts'
        );

        // Lint the code whenever the model is created or changed
        const handleLinting = async (model) => {
            const code = model.getValue();

            try {
                const lintMessages = await eslintInstance.lintText(code, {
                    filePath: model.uri.path,
                });

                // Convert ESLint results to Monaco markers
                const markers = lintMessages[0].messages.map((msg) => ({
                    startLineNumber: msg.line,
                    startColumn: msg.column,
                    endLineNumber: msg.endLine || msg.line,
                    endColumn: msg.endColumn || msg.column,
                    message: msg.message,
                    severity: monaco.MarkerSeverity.Error, // You can change severity based on the rule
                }));

                // Set markers in Monaco editor
                monaco.editor.setModelMarkers(model, 'eslint', markers);
            } catch (error) {
                console.error('Error during linting:', error);
            }
        };

        // Lint when the editor's model changes
        editor.onDidChangeModelContent(() => {
            const model = editor.getModel();
            if (model) {
                handleLinting(model);
            }
        });

        // Initial lint on model creation
        monaco.editor.onDidCreateModel((model) => {
            handleLinting(model);
        });
    };

    return (
        <Editor
            height="500px"
            defaultLanguage="typescript"
            defaultValue="// type your code here"
            onMount={onEditorMount}
            options={{
                automaticLayout: true,
                minimap: { enabled: false },
            }}
        />
    );
}

export default CodeEditor;`

**Comment by Rahul5546-X** [comment]
@restarajat 

Most of  your code is correct, there are just some minor changes required.

Instead of es-lint, you should user es-lint browserify.

import * as eslint from 'eslint-linter-browserify';

       const esLintRules = {
	      'capitalized-comments': 'warn',
	      'comma-dangle': [
		    'error',
		   'never'
	        ],
	    'no-unused-vars': 'warn'
      };

1. Remove your eslintInstance

2. In the try-catch block

    a. Update this
                                // const lintMessages = await eslintInstance.lintText(code, {
				           // 	filePath: model.uri.path
       				// });

				const lintMessages = new eslint.Linter().verify(code, [{
					rules: esLintRules
				}]);
				
                                and this
                               //   const markers = lintMessages[0].messages.map((msg) => ({
			       const markers = lintMessages.map((msg) => ({

**Reply by Rahul5546-X** [reply]
If you want the feature to apply quick fixes in your editor. Register a codeActionProvider, I can also help you with that. If required feel free to ask
